File: src/command/workflow_live_generated_contract_retry.js
  function generatedContractFocusedVerificationItem(item) {
    return item && generatedContractVerificationIntent(item)
      && generatedContractPresent(item.focused_verification)
      && (item.expected_evidence !== undefined || item.artifact_requirements !== undefined)
      && !generatedContractPresent(item.target_files);
  }

  function generatedContractVerificationIntent(item) {
    const workType = String((item && (item.work_type || item.workType)) || "").toLowerCase();
    if (!workType || workType.includes("verification")) return true;
    return workType === "verified_noop" && generatedContractRetryVerificationIntent(item);
  }

  function generatedContractRetryVerificationIntent(item) {
    return generatedContractPresent(item && (item.retry_command_shape || item.retryCommandShape))
      || generatedContractPresent(item && (item.retry_steps || item.retrySteps))
      || generatedContractPresent(item && (item.expected_result_shape || item.expectedResultShape))
      || generatedContractPresent(item && (item.retry_plan || item.retryPlan));
  }

  function generatedContractEmbeddedTaskIds(source) {
    const haystack = generatedContractRawStrings(source, ["item_id", "id", "source_item_id", "sourceItemId", "failed_item_id", "failedItemId", "source_failed_item_id", "sourceFailedItemId"]).join(" ").toLowerCase();
    return generatedContractUnique(generatedContractArray(taskUniverse.tasks || []).map((task) => task && task.canonical_task_id).filter((id) => {
      return id && generatedContractEmbeddedTaskCandidates(String(id)).some((candidate) => haystack.includes(candidate));
    }));
  }

  function generatedContractEmbeddedTaskCandidates(canonical) {
    const parts = String(canonical).split("-");
    const candidates = [String(canonical).toLowerCase()];
    if (parts.length > 2) candidates.push(parts.slice(1).join("-").toLowerCase());
    const digits = parts[parts.length - 1] || "";
    if (/^[0-9]+$/.test(digits)) candidates.push(("T" + digits).toLowerCase());
    return generatedContractUnique(candidates);
  }

  function generatedContractInventoryRoots(source) {
    const roots = [source].filter(Boolean);
    for (const path of [["data"], ["result"], ["result", "data"], ["result", "result"], ["result", "result", "data"], ["data", "result"], ["data", "result", "data"]]) {
      const root = path.reduce((current, key) => current && current[key], source);
      if (root && typeof root === "object") roots.push(root);
    }
    return roots;
  }

  function generatedContractInventoryRootItems(root) {
    const dataItems = root && root.items;
    return generatedContractArray(root && root.items)
      .concat(generatedContractArray(root && root.inventory && root.inventory.items))
      .concat(generatedContractArray(root && root.repaired_items))
      .concat(generatedContractArray(root && root.implementation_items))
      .concat(generatedContractArray(root && root.verified_noop_items))
      .concat(generatedContractArray(dataItems && dataItems.repaired_items))
      .concat(generatedContractArray(dataItems && dataItems.implementation_items))
      .concat(generatedContractArray(dataItems && dataItems.verified_noop_items));
  }

  function generatedContractInventorySourceItems(source) {
    if (Array.isArray(source && source.items)) {
      return source.items;
    }
    return generatedContractInventoryRoots(source)
      .flatMap((root) => generatedContractInventoryRootItems(root))
      .concat(generatedContractInventoryRetryItems(source));
  }

  function generatedContractInventorySourceIssues(source) {
    if (Array.isArray(source && source.items)) {
      return generatedContractArray(source.unresolved_issues);
    }
    return generatedContractInventoryRoots(source)
      .flatMap((root) => generatedContractArray(root && root.unresolved_issues));
  }

  function generatedContractRetryValues(source, keys) {
    return generatedContractRawValues(source, ["retry_steps", "retrySteps", "retry_plan", "retryPlan"])
      .flatMap((step) => generatedContractRawValues(step, keys));
  }

  function generatedContractInventoryRetryItems(source) {
    return generatedContractInventoryRoots(source).flatMap((root) => {
      const plan = root.repair_plan || root.repairPlan || {};
      return generatedContractRawValues(root, ["retry_items", "retryItems"])
        .concat(generatedContractRawValues(plan, ["retry_items", "retryItems"]));
    });
  }

  function generatedContractApplyRetryAliases(source, out) {
    const manualRetry = source.manual_fixture_retry || source.manualFixtureRetry || {};
    const commandEntry = source.commands_run_entry || source.commandsRunEntry || {};
    const completionEvidence = source.required_completion_evidence || source.requiredCompletionEvidence || source.completion_evidence_shape || source.completionEvidenceShape || {};
    const retryCommands = generatedContractRawValues(manualRetry, ["commands", "commands_run", "commandsRun"])
      .concat(generatedContractRawValues(source, ["retry_command", "retryCommand", "retry_commands", "retryCommands"]))
      .concat(generatedContractNestedAliasValues(source, ["retry_command_shape", "retryCommandShape"], ["command", "commands"]))
      .concat(generatedContractRawValues(commandEntry, ["command", "commands"]))
      .concat(generatedContractRawValues(completionEvidence, ["command_refs", "commandRefs"]))
      .concat(generatedContractRetryValues(source, ["command", "commands", "commands_run", "commandsRun"]));
    if (retryCommands.length > 0) {
      out.focused_verification = generatedContractArray(out.focused_verification).concat(retryCommands);
    }
    const retryExpected = generatedContractRawValues(source, ["acceptance_rule", "acceptanceRule"])
      .concat(generatedContractRawValues(completionEvidence, ["evidence_refs", "evidenceRefs", "required_summary_points", "requiredSummaryPoints"]))
      .concat(generatedContractRetryValues(source, ["required_evidence", "requiredEvidence", "evidence_to_capture", "evidenceToCapture", "expected_result", "expectedResult"]));
    if (retryExpected.length > 0) {
      out.expected_evidence = generatedContractArray(out.expected_evidence).concat(retryExpected);
    }
    const retryArtifacts = generatedContractRawValues(manualRetry, ["artifact_checks", "artifactChecks", "expected_artifacts", "expectedArtifacts", "artifact_paths", "artifactPaths"])
      .concat(generatedContractRawValues(completionEvidence, ["artifact_paths", "artifactPaths"]))
      .concat(generatedContractRetryValues(source, ["artifact_checks", "artifactChecks", "expected_artifacts", "expectedArtifacts", "artifact_paths", "artifactPaths"]));
    if (retryArtifacts.length > 0) {
      out.artifact_requirements = generatedContractArray(out.artifact_requirements).concat(retryArtifacts);
    }
    if (!out.source_item_id && (source.source_failed_item_id || source.sourceFailedItemId)) {
      out.source_item_id = source.source_failed_item_id || source.sourceFailedItemId;
    }
  }

  function generatedContractNestedAliasValues(source, roots, keys) {
    return roots.flatMap((root) => generatedContractArrayOrOne(source && source[root]))
      .flatMap((entry) => generatedContractRawValues(entry, keys));
  }

  function generatedContractArrayOrOne(value) {
    if (Array.isArray(value)) return value;
    return value === undefined || value === null ? [] : [value];
  }

  function generatedContractApplyProviderEnvAliases(source, out) {
    const keys = generatedContractProviderEnvKeys(source);
    if (keys.length > 0) {
      out.provider_env_requirements = generatedContractArray(out.provider_env_requirements).concat(keys);
      out.expected_evidence = generatedContractArray(out.expected_evidence).concat(keys.flatMap((key) => generatedContractStrings(key).map((name) => "provider_env_proof:" + name)));
    }
    if (!out.provider_env_proof && (source.provider_env_proof || source.providerEnvProof)) {
      out.provider_env_proof = source.provider_env_proof || source.providerEnvProof;
    }
  }

  function generatedContractProviderEnvKeys(source) {
    return generatedContractUnique(generatedContractRawValues(source, ["provider_env_requirements", "providerEnvRequirements", "provider_env_required_keys", "providerEnvRequiredKeys", "required_env_keys", "requiredEnvKeys", "credential_env_keys", "credentialEnvKeys"])
      .flatMap((key) => generatedContractStrings(key))
      .map((key) => String(key).trim().toUpperCase())
      .filter(Boolean));
  }

File: src/command/workflow_live_generated_contract_retry.rs
fn normalize_retry_context(
    value: &serde_json::Value,
    object: &mut serde_json::Map<String, serde_json::Value>,
) {
    let mut commands = retry_step_values(
        value,
        &["command", "commands", "commands_run", "commandsRun"],
    );
    commands.extend(raw_values_from_aliases(
        value,
        &[
            "retry_command",
            "retryCommand",
            "retry_commands",
            "retryCommands",
            "recommended_retry",
            "recommendedRetry",
            "recommended_retries",
            "recommendedRetries",
        ],
    ));
    commands.extend(nested_alias_values(
        value,
        &["retry_command_shape", "retryCommandShape"],
        &["command", "commands"],
    ));
    commands.extend(nested_alias_values(
        value,
        &["commands_run_entry", "commandsRunEntry"],
        &["command", "commands"],
    ));
    commands.extend(nested_alias_values(
        value,
        &completion_evidence_roots(),
        &["command_refs", "commandRefs"],
    ));
    append_alias_values(object, "focused_verification", commands);
    let mut expected = retry_step_values(
        value,
        &[
            "required_evidence",
            "requiredEvidence",
            "evidence_to_capture",
            "evidenceToCapture",
            "expected_result",
            "expectedResult",
        ],
    );
    expected.extend(raw_values_from_aliases(
        value,
        &["acceptance_rule", "acceptanceRule"],
    ));
    if expected.is_empty() {
        expected.extend(raw_values_from_aliases(
            value,
            &["recommended_retry", "recommendedRetry"],
        ));
    }
    expected.extend(nested_alias_values(
        value,
        &completion_evidence_roots(),
        &[
            "evidence_refs",
            "evidenceRefs",
            "required_summary_points",
            "requiredSummaryPoints",
        ],
    ));
    append_alias_values(object, "expected_evidence", expected);
    let artifacts = nested_alias_values(
        value,
        &completion_evidence_roots(),
        &["artifact_paths", "artifactPaths"],
    );
    append_artifact_requirement_values(object, artifacts, false);
    if !object.contains_key("source_item_id")
        && let Some(source_item_id) =
            first_string(value, &["source_failed_item_id", "sourceFailedItemId"])
        {
            object.insert(
                "source_item_id".to_string(),
                serde_json::Value::String(source_item_id),
            );
        }
}

fn normalize_provider_env_context(
    value: &serde_json::Value,
    object: &mut serde_json::Map<String, serde_json::Value>,
) {
    let requirements = provider_env_requirement_values(value);
    append_alias_values(object, "provider_env_requirements", requirements.clone());
    append_alias_values(
        object,
        "expected_evidence",
        provider_env_expected_evidence(requirements),
    );
    copy_alias_value(
        value,
        &["provider_env_proof", "providerEnvProof"],
        object,
        "provider_env_proof",
    );
}

fn provider_env_requirement_aliases() -> [&'static str; 8] {
    [
        "provider_env_requirements",
        "providerEnvRequirements",
        "provider_env_required_keys",
        "providerEnvRequiredKeys",
        "required_env_keys",
        "requiredEnvKeys",
        "credential_env_keys",
        "credentialEnvKeys",
    ]
}

fn provider_env_expected_evidence(values: Vec<serde_json::Value>) -> Vec<serde_json::Value> {
    values
        .into_iter()
        .flat_map(value_to_strings)
        .map(|key| serde_json::json!(format!("provider_env_proof:{key}")))
        .collect()
}

fn provider_env_requirement_values(value: &serde_json::Value) -> Vec<serde_json::Value> {
    raw_values_from_aliases(value, &provider_env_requirement_aliases())
        .into_iter()
        .flat_map(value_to_strings)
        .map(|key| key.trim().to_ascii_uppercase())
        .filter(|key| !key.is_empty())
        .collect::<BTreeSet<_>>()
        .into_iter()
        .map(serde_json::Value::String)
        .collect()
}

fn collect_generated_inventory_items(value: &serde_json::Value) -> Vec<serde_json::Value> {
    if let Some(items) = top_level_items(value) {
        return items.clone();
    }
    let mut out = Vec::new();
    for root in generated_inventory_roots(value) {
        push_inventory_root_items(root, &mut out);
    }
    push_generated_retry_plan_items(value, &mut out);
    out
}

fn top_level_items(value: &serde_json::Value) -> Option<&Vec<serde_json::Value>> {
    value
        .get("items")
        .and_then(serde_json::Value::as_array)
}

fn generated_inventory_roots(value: &serde_json::Value) -> Vec<&serde_json::Value> {
    let mut roots = vec![value];
    for path in [
        &["data"][..],
        &["result"][..],
        &["result", "data"][..],
        &["result", "result"][..],
        &["result", "result", "data"][..],
        &["data", "result"][..],
        &["data", "result", "data"][..],
    ] {
        if let Some(root) = value_at_path(value, path) {
            roots.push(root);
        }
    }
    roots
}

fn value_at_path<'a>(
    value: &'a serde_json::Value,
    path: &[&str],
) -> Option<&'a serde_json::Value> {
    path.iter().try_fold(value, |current, key| current.get(*key))
}

fn push_inventory_root_items(root: &serde_json::Value, out: &mut Vec<serde_json::Value>) {
    push_item_collection(root.get("items"), out);
    push_item_collection(root.get("inventory").and_then(|v| v.get("items")), out);
    push_array(root.get("repaired_items"), out);
    push_array(root.get("implementation_items"), out);
    push_array(root.get("verified_noop_items"), out);
    if let Some(items) = root.get("items") {
        push_array(items.get("repaired_items"), out);
        push_array(items.get("implementation_items"), out);
        push_array(items.get("verified_noop_items"), out);
    }
}

fn push_item_collection(value: Option<&serde_json::Value>, out: &mut Vec<serde_json::Value>) {
    match value {
        Some(serde_json::Value::Array(items)) => out.extend(items.iter().cloned()),
        Some(serde_json::Value::Object(items)) => push_grouped_item_buckets(items, out),
        _ => {}
    }
}

fn push_grouped_item_buckets(
    items: &serde_json::Map<String, serde_json::Value>,
    out: &mut Vec<serde_json::Value>,
) {
    for key in [
        "implementation",
        "implementations",
        "implementation_items",
        "remediation",
        "remediation_items",
        "review_remediation",
        "review_remediation_items",
        "focused_verification",
        "focused_verification_items",
        "review_verification",
        "review_verification_items",
        "verified_noop",
        "verified_noops",
        "verified_noop_items",
    ] {
        push_array(items.get(key), out);
    }
}

fn push_array(value: Option<&serde_json::Value>, out: &mut Vec<serde_json::Value>) {
    if let Some(items) = value.and_then(serde_json::Value::as_array) {
        out.extend(items.iter().cloned());
    }
}

fn completion_evidence_roots() -> [&'static str; 4] {
    [
        "required_completion_evidence",
        "requiredCompletionEvidence",
        "completion_evidence_shape",
        "completionEvidenceShape",
    ]
}

fn push_generated_retry_plan_items(value: &serde_json::Value, out: &mut Vec<serde_json::Value>) {
    for root in generated_inventory_roots(value) {
        push_array(root.get("retry_items"), out);
        push_array(root.get("retryItems"), out);
        if let Some(plan) = root.get("repair_plan").or_else(|| root.get("repairPlan")) {
            push_array(plan.get("retry_items"), out);
            push_array(plan.get("retryItems"), out);
        }
    }
}

fn retry_step_values(value: &serde_json::Value, aliases: &[&str]) -> Vec<serde_json::Value> {
    ["retry_steps", "retrySteps", "retry_plan", "retryPlan"]
        .iter()
        .filter_map(|key| value.get(*key))
        .flat_map(|steps| match steps {
            serde_json::Value::Array(items) => items.clone(),
            serde_json::Value::Null => Vec::new(),
            other => vec![other.clone()],
        })
        .flat_map(|step| raw_values_from_aliases(&step, aliases))
        .collect()
}

fn nested_alias_values(
    value: &serde_json::Value,
    roots: &[&str],
    aliases: &[&str],
) -> Vec<serde_json::Value> {
    roots
        .iter()
        .filter_map(|root| value.get(*root))
        .flat_map(|root| match root {
            serde_json::Value::Array(items) => items
                .iter()
                .flat_map(|item| raw_values_from_aliases(item, aliases))
                .collect::<Vec<_>>(),
            _ => raw_values_from_aliases(root, aliases),
        })
        .collect()
}
